File: sert-2014/src/edu/wpi/first/wpilibj/templates/subsystems/DriveSubsystem.java

package edu.wpi.first.wpilibj.templates.subsystems;

import edu.wpi.first.wpilibj.templates.CANJaguar;
import edu.wpi.first.wpilibj.can.CANTimeoutException;
import edu.wpi.first.wpilibj.command.Subsystem;
import edu.wpi.first.wpilibj.templates.CANJaguarMaster;
import edu.wpi.first.wpilibj.templates.CANRobotDrive;
import edu.wpi.first.wpilibj.templates.OI;
import edu.wpi.first.wpilibj.templates.RobotMap;
import edu.wpi.first.wpilibj.templates.commands.TeleoperatedDrive;

/**
 *
 */
public class DriveSubsystem extends Subsystem {
    CANRobotDrive drive;
    CANJaguarMaster rightSide;
    private boolean isArcade = true;
    private boolean slowMode = false;
    
    public DriveSubsystem() {
        try{
            CANJaguar right_rear = new CANJaguar(RobotMap.RIGHT_REAR_DRIVE_JAG, CANJaguar.ControlMode.kVoltage);
            CANJaguar left_rear = new CANJaguar(RobotMap.LEFT_REAR_DRIVE_JAG, CANJaguar.ControlMode.kVoltage);
            
            CANJaguarMaster left = new CANJaguarMaster(RobotMap.LEFT_FRONT_DRIVE_JAG, left_rear);
            CANJaguarMaster right = new CANJaguarMaster(RobotMap.RIGHT_FRONT_DRIVE_JAG, right_rear);
            drive = new CANRobotDrive(left, right);
        } catch (CANTimeoutException ex) {
            ex.printStackTrace();
        }

    }

    public void resetCANBus() throws CANTimeoutException {
        drive.changeControlMode(CANJaguar.ControlMode.kPosition);
        drive.configEncoderCodesPerRev(360); 
        drive.setPositionReference(CANJaguar.PositionReference.kQuadEncoder);
        drive.setPID(RobotMap.K_P, RobotMap.K_I, RobotMap.K_D);
        drive.changeControlMode(CANJaguar.ControlMode.kPercentVbus);
        //drive.setVoltageRampRate(60);
    }
    
    public void teleoperatedDrive() {
        if (isArcade) {
            arcade();
        } else {
            tank();
        }
    }
    
    public void changeControlMode(CANJaguar.ControlMode mode) {
        try {
            drive.changeControlMode(mode);
        } catch (CANTimeoutException ex) {
//            ex.printStackTrace();
            ex.getMessage();
        }
    }
    
    public void moveToPosition(double position) {
        try {
            drive.moveToPosition(position);
        } catch (CANTimeoutException ex) {
//            ex.printStackTrace();
        }
        
    }
    
    public void enableControl() {
        try {
            drive.enableControl();
        } catch (CANTimeoutException ex) {
//            ex.printStackTrace();
        }
    }
    
    public void enableControl(double position) {
        try {
            drive.enableControl(position);
        } catch (CANTimeoutException ex) {
//            ex.printStackTrace();
        }
    }
    
    public void disableControl() {
        try {
            drive.disableControl();
        } catch (CANTimeoutException ex) {
//            ex.printStackTrace();
        }
    }
    
    public double getPosition() {
        try {
            return drive.getPosition();
        } catch (CANTimeoutException ex) {
//            ex.printStackTrace();
        }
        return 0;
    }

    public void changeTeleoperatedDriveMode() {
        isArcade = !isArcade;
    }
    
    public double getOutputVoltage() {
        try {
            return drive.getOutputVoltage();
        } catch (CANTimeoutException ex) {
//            ex.printStackTrace();
        }
        return 0;
    }
   
    private void arcade(){
        if (slowMode) {
            drive.arcadeDrive(-OI.getInstance().getLeftDriveStick().getY()*.6, -OI.getInstance().getLeftDriveStick().getX()*.6, true);
        } else {
            drive.arcadeDrive(-OI.getInstance().getLeftDriveStick().getY(), -OI.getInstance().getLeftDriveStick().getX(), true);
        }
    }
    
    private void tank() {
        if (slowMode) {
            drive.tankDrive(-OI.getInstance().getLeftDriveStick().getY()*.3, -OI.getInstance().getRightDriveStick().getY()*.3, true);
        } else {
            drive.tankDrive(-OI.getInstance().getLeftDriveStick().getY(), -OI.getInstance().getRightDriveStick().getY(), true);
        }
    }
    
    public void slowMode() {
        slowMode = !slowMode;
    }
    
    public void initDefaultCommand() {
       setDefaultCommand(new TeleoperatedDrive());
    }

    public double getRightPosition() throws CANTimeoutException {
        return drive.getRightPosition();
    }

}

File: sert-2014-C++/Subsystems/DriveSubsystem.cpp
// swapping ostringstream objects
#include <string>       // std::string
#include <iostream>     // std::cout
#include <sstream>      // std::stringstream

#include "DriveSubsystem.h"
#include "../RobotMap.h"
#include "../OI.h"
#include "../Commands/TeleoperatedDrive.h"

DriveSubsystem::DriveSubsystem() :
	Subsystem( "DriveSubsystem" )
{
	try
	{
	   m_drive = new CANRobotDrive( new CANJaguarMaster( LEFT_FRONT_DRIVE_JAG, new CANJaguar(LEFT_REAR_DRIVE_JAG)), new CANJaguarMaster(RIGHT_FRONT_DRIVE_JAG, new CANJaguar(RIGHT_REAR_JAG_DRIVE)));
	} 
	catch ( int ex ) 
	{
		std::cerr << "An exception occurred. Exception Nr. " << ex << '\n';
	}
}

void DriveSubsystem::SetTeleoperatedDrive() 
{
	if ( m_isArcade ) 
	{
		Arcade();
	} 
	else 
	{
		Tank();
	}
	PrintEncoders();
}

void DriveSubsystem::ChangeControlMode( CANJaguar::ControlMode mode ) 
{
	try 
	{
		m_drive->ChangeControlMode( mode );
	} 
	catch ( int ex ) 
	{
		std::cerr << "An exception occurred. Exception Nr. " << ex << '\n';
	}
}

void DriveSubsystem::MoveDistance( double inches ) 
{
	try 
	{
		m_drive->MoveInches( inches );
	} 
	catch ( int ex ) 
	{
		std::cerr << "An exception occurred. Exception Nr. " << ex << '\n';
	}
}

string int2str(int inp) { stringstream str; str << inp; return str.str(); }
string double2str(double in) { stringstream str; str << in; return str.str(); }

void DriveSubsystem::PrintEncoders() 
{
	try 
	{
		DriverStationLCD::GetInstance()->PrintfLine( DriverStationLCD::kUser_Line1, double2str( m_drive->GetEncoders() ).c_str() );
		DriverStationLCD::GetInstance()->UpdateLCD();
	} 
	catch ( int ex ) 
	{
		std::cerr << "An exception occurred. Exception Nr. " << ex << '\n';
	}
}

void DriveSubsystem::ChangeTeleoperatedDriveMode() 
{
	m_isArcade = !m_isArcade;
}

void DriveSubsystem::Arcade()
{
	if ( m_slowMode ) 
	{
		m_drive->ArcadeDrive(-OI::GetInstance()->GetLeftDriveStick()->GetY()*.3, -OI::GetInstance()->GetLeftDriveStick()->GetX()*.3, true);
	} 
	else 
	{
		m_drive->ArcadeDrive(-OI::GetInstance()->GetLeftDriveStick()->GetY(), -OI::GetInstance()->GetLeftDriveStick()->GetX(), true);
	}
}

void DriveSubsystem::Tank() 
{
	if ( m_slowMode ) 
	{
		m_drive->TankDrive(-OI::GetInstance()->GetLeftDriveStick()->GetY()*.3, -OI::GetInstance()->GetRightDriveStick()->GetY()*.3, true);
	} 
	else 
	{
		m_drive->TankDrive(-OI::GetInstance()->GetLeftDriveStick()->GetY(), -OI::GetInstance()->GetRightDriveStick()->GetY(), true);
	}
}

void DriveSubsystem::SlowMode() 
{
	m_slowMode = true;
}

void DriveSubsystem::InitDefaultCommand() 
{
	SetDefaultCommand( new TeleoperatedDrive() );
}
